geoip-shell: add package

friendly-bits opened the pull request:
Maintainer: Anton Khazan / antonk / @friendly-bits
Compile tested: x86-64 v23.05.3
Run tested: x86-64 VM with OpenWrt v23.05.2, WNDR3700 v2 with OpenWrt SFE r15568-b3eccbca7c (approximately 2021 release)

Description:
Flexible geoip blocker with user-friendly command-line interface

This is my first pull request, sorry in advance if something is wrong with it.

Please note that I'm stripping comments from the code in this package to reduce its size. You can read the code with comments in the project's repository.

friendly-bits commented:
Please let me know if I need to do something about formalities... When pushing the merge with latest changes, I of course forgot to signoff, but I later did `git rebase --signoff HEAD~1`, so it should be good now (?)

friendly-bits commented:
I'm not sure what to do about formalities. I don't really want to put my official name on the Github account, but then CONTRIBUTING.md says that I need so signoff with my official name an email address that is not the github address. So I added that info to the initial pull request, and then signed off all further commits with it. This is, however, failing the formalities check. Please let me know if I need to do something about this.
@hnyman tagging you because you initiated the checks.

robimarko commented:
Formalities check does not check against your GH account, it only checks whether the commit author matches the one in Signed-off-by.

Also, please get rid of all of the merge commits, and add the required prefix as well.

friendly-bits commented:
> Formalities check does not check against your GH account, it only checks whether the commit author matches the one in Signed-off-by.
> 
> Also, please get rid of all of the merge commits, and add the required prefix as well.

Thanks. I did git rebase out of Stan's commit just before the `add geoip-shell` commit, changing the Author of the `add geoip-shell` commit in order for it to match the signed-off-by, and changing the commit subjects of all my following commits. Is it good now or do I need to make additional changes?

robimarko commented:
Wait, did you use somebody elses work as a base?

If so, then the original authorship needs to be preserved by keeping the author field and first Signed-off-by to the original author and you just add your own signed-off-by at the bottom.

friendly-bits commented:
> Wait, did you use somebody elses work as a base?

No. I followed this guide to retroactively change the commits, and it explains that to do that you need to pick a commit just before the first commit you want to change... So the commit just before my commits happened to be Stan's commit.
https://stackoverflow.com/questions/3042437/how-can-i-change-the-commit-author-for-a-single-commit

robimarko commented:
Well, something went wrong, can you please just rebase your branch on top of the master?

And add commit descriptions for each commit as you dont have any

friendly-bits commented:
> Well, something went wrong, can you please just rebase your branch on top of the master?
> 
> And add commit descriptions for each commit as you dont have any

Done

robimarko commented:
Well, considering you are adding a brand new package there is no need for multiple commits, just squash them.

And have one uniform commit message describing what you are adding and why.

Also, why are you including the "code" locally, if there is an upstream repo use that.

friendly-bits commented:
> Also, why are you including the "code" locally, if there is an upstream repo use that.

There is an upstream repo but the code there is not suited for an OpenWrt package. I wrote a script which basically generates everything needed for OpenWrt. I'm not sure how to handle this situation without adding the code locally.

friendly-bits commented:
> Well, considering you are adding a brand new package there is no need for multiple commits, just squash them.
> 
> And have one uniform commit message describing what you are adding and why.

Done

friendly-bits commented:
@robimarko big thank you for your help. Please let me know if there is anything else I need to do, or if we're done and I can start looking for reviewers.

robimarko commented:
You still have multiple commits, squash them.

I dont like carrying the source in OpenWrt repos, that should be the last resort.
Why not have a dedicated release GH repo?

friendly-bits commented:
> Why not have a dedicated release GH repo?

I can do that but then I'll need to cancel this pull request and add a new one?

robimarko commented:
> > Why not have a dedicated release GH repo?
> 
> I can do that but then I'll need to cancel this pull request and add a new one?

No, simply make the changes and force push them, no reason for new PR

friendly-bits commented:
@robimarko It'll take me some time to figure out the details of how to automate releases upload to a new repository and change the Makefile. I'll let you know once this is done.

friendly-bits commented:
@robimarko An additional question: what to do about the documentation? I'd like to link to it from the description of the package. I could link to my own repository but I assume you would prefer the link to be guaranteed to always work, which means it's gotta be in the OpenWrt repo?

friendly-bits commented:
@robimarko

> No, simply make the changes and force push them, no reason for new PR

Done. I hope I did this correctly.

> You still have multiple commits, squash them.

Also done. Squashed everything into "geoip-shell: add package". Not sure what to write in the description, so I left it empty.

As to documentation, I added it along with the Makefile, please let me know if I need to change this.

friendly-bits commented:
@robimarko Done

friendly-bits commented:
The above force pushes improve the documentation and convert documentation markdown links to relative, so the links should work in the OpenWrt repo.

friendly-bits commented:
Sorry about the last-minute pushes, this should be the last one.

robimarko commented:
Thanks, I have rebased this to let CI do its thing.

Hopefully some of the package maintainers will give it a look.

brada4 reviewed (changes requested):
Request change:
- add random sleep in crontab script for like half of period.

Summary:
Looks nicely integrated, few points to consider in future devlopment (I shared in forum)
- make independent hook right before raw/prerouting to save slab allocations for new states thus using less resources dropping same packet also blocking out existing connections on change
- same but deleting conntrack states that fall in blacklist with conntrack command if such is installed
- in absence of nftables or selectable way to change overall engine to `ip route blackhole`

brada4 reviewed (changes requested):
Can it persist geoip data cf https://github.com/pvxe/nftables-geoip referenced from nftables documentation?

friendly-bits commented:
> Can it persist geoip data cf https://github.com/pvxe/nftables-geoip referenced from nftables documentation?

What means 'cf' and what do you mean by 'persist'? Please link to the documentation...
If you mean generating mappings between countries and marks (which can be later used in user-defined rules) then geoip-shell does not do that. Should it?

brada4 commented:
Kind of existing framework is not broadly instrumented... Yours achieves result but will squash if upstream comes to compete.

friendly-bits commented:
> Kind of existing framework is not broadly instrumented... Yours achieves result but will squash if upstream comes to compete.

I know that you have good intentions but understanding you is very challenging for me. And you are not answering clarification questions, or answering in a vague way. Please put more effort into communicating your ideas clearly.

friendly-bits commented:
> Request change:
> 
> * add random sleep in crontab script for like half of period.

Proposed code change in the -run script which implements sleeping a random number of seconds (0-300) before executing the 'update' action when called from the cron job.

```
# wait $reboot_sleep seconds after boot, or 0-300 seconds before updating
[ "$daemon_mode" ] && {
	if [ "$action_run" = restore ]; then
		IFS='. ' read -r uptime _ < /proc/uptime
		: "${uptime:=0}"
		: "${reboot_sleep:=30}"
		sl_time=$((reboot_sleep-uptime))
	elif [ "$action_run" = update ]; then
		rand_int="$(tr -cd 0-9 < /dev/urandom | dd bs=1 count=3 2>/dev/null)"
		: "${rand_int:=0}"
		sl_time=$((rand_int*300/999))
	fi
	[ $sl_time -gt 0 ] && {
		echolog "Sleeping for ${sl_time}s..."
		sleep $sl_time
	}
}

```

hnyman commented on net/geoip-shell/Makefile:
```diff
@@ -0,0 +1,162 @@
+
+PKG_NAME:=geoip-shell
+PKG_VERSION:=0.5
+PKG_RELEASE:=r2
```
The PKG_RELEASE has the extra "r" in its release, leading to extra r also in the final filename:
`geoip-shell_0.5-rr2_all.ipk`
( https://downloads.openwrt.org/snapshots/packages/arm_cortex-a9_neon/packages/ )

The r is automatically added by the build system.

I will remove it.

friendly-bits replied:
@hnyman That's odd - When building on my machine, the filename comes out correct. Should I remove the 'r' in the Makefile in future releases? So it'll be like `PKG_RELEASE:=1` ?

hnyman replied:
> When building on my machine, the filename comes out correct. 

Are you sure that you are compiling it with really up-to-date OpenWrt master ?
The logic regarding "r" addition was added just a few weeks ago. If your build env does not  have up-to-date code, it won't have the logic.

(the whole 'r' there is quite new)

friendly-bits replied:
> Are you sure that you are compiling it with really up-to-date OpenWrt master ?

Ah yes, pulled latest master and now filename has 2 r's.
